source/extensions/filters/http/basic_auth/config.cc

```diff
@@ -12,36 +12,50 @@ using envoy::extensions::filters::http::basic_auth::v3::BasicAuth;
 
 namespace {
 
-UserMapConstPtr readHtpasswd(const std::string& htpasswd) {
-  std::unique_ptr<absl::flat_hash_map<std::string, User>> users =
-      std::make_unique<absl::flat_hash_map<std::string, User>>();
+UserMap readHtpasswd(const std::string& htpasswd) {
+  UserMap users;
+
   std::istringstream htpsswd_ss(htpasswd);
   std::string line;
 
   while (std::getline(htpsswd_ss, line)) {
+    // TODO(wbpcode): should we trim the spaces or empty chars?
+
+    // Skip empty lines and comments.
+    if (line.empty() || line[0] == '#') {
+      continue;
+    }
+
     const size_t colon_pos = line.find(':');
+    if (colon_pos == std::string::npos) {
+      throw EnvoyException("basic auth: invalid htpasswd format, username:password is expected");
+    }
 
-    if (colon_pos != std::string::npos) {
-      std::string name = line.substr(0, colon_pos);
-      std::string hash = line.substr(colon_pos + 1);
+    std::string name = line.substr(0, colon_pos);
+    std::string hash = line.substr(colon_pos + 1);
 
-      if (name.empty()) {
-        throw EnvoyException("basic auth: invalid user name");
-      }
+    if (name.empty() || hash.empty()) {
+      throw EnvoyException("basic auth: empty user name or password");
+    }
+
+    // TODO(wbpcode): user name in exception message would simplify the debugging. But if we think
+    // it also sensitive information, we can remove it.
+    if (users.contains(name)) {
+      throw EnvoyException(fmt::format("basic auth: duplicate user '{}'", name));
+    }
 
-      if (absl::StartsWith(hash, "{SHA}")) {
-        hash = hash.substr(5);
-        // The base64 encoded SHA1 hash is 28 bytes long
-        if (hash.length() != 28) {
-          throw EnvoyException("basic auth: invalid SHA hash length");
-        }
+    if (!absl::StartsWith(hash, "{SHA}")) {
+      throw EnvoyException(fmt::format(
+          "basic auth: unsupported htpasswd format: please use {{SHA}} for '{}'", name));
+    }
 
-        users->insert({name, {name, hash}});
-        continue;
-      }
+    hash = hash.substr(5);
+    // The base64 encoded SHA1 hash is 28 bytes long
+    if (hash.length() != 28) {
+      throw EnvoyException(fmt::format("basic auth: invalid SHA hash length for '{}'", name));
     }
 
-    throw EnvoyException("basic auth: unsupported htpasswd format: please use {SHA}");
+    users.insert({name, {name, hash}});
```

Comment:
- handle empty lines and comment lines.
- handle duplicate users.
- handle empty hash.
- tread no colon and non SHA as different format error.